File: src/main/java/com/protryon/jasm/instruction/psuedoinstructions/EnterTry.java
package com.protryon.jasm.instruction.psuedoinstructions;

import com.protryon.jasm.Constant;
import com.protryon.jasm.JType;
import com.protryon.jasm.Method;
import com.protryon.jasm.instruction.Instruction;
import com.shapesecurity.functional.F;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;

public class EnterTry extends Instruction {

    public Label catchBlock;
    public JType exceptionType;

    public EnterTry(Label catchBlock, JType exceptionType) {
        this.catchBlock = catchBlock;
        this.exceptionType = exceptionType;
    }

    @Override
    public String name() {
        return "TRY";
    }

    @Override
    public int opcode() {
        return -1;
    }

    @Override
    public void read(boolean wide, ArrayList<Constant> constants, Method method, F<Integer, Label> labelMaker, int pc, DataInputStream in) throws IOException {
        throw new UnsupportedOperationException();
    }

    @Override
    public void write(boolean wide, DataOutputStream out, F<Label, Integer> labelIndexer, F<Constant, Integer> constantIndexer, int pc) throws IOException {
        throw new UnsupportedOperationException();
    }

    @Override
    public int pushes() {
        return 0;
    }

    @Override
    public int pops() {
        return 0;
    }

    @Override
    public String toString() {
        return "TRY " + exceptionType.niceName + " -> " + catchBlock.toString();
    }

    @Override
    public Instruction fromString(String str) {
        return null;
    }
}

File: src/main/java/com/protryon/jasm/instruction/psuedoinstructions/EnterTry.kt
package com.protryon.jasm.instruction.psuedoinstructions

import com.protryon.jasm.Constant
import com.protryon.jasm.JType
import com.protryon.jasm.Method
import com.protryon.jasm.instruction.Instruction
import com.shapesecurity.functional.F

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.util.ArrayList

class EnterTry(var catchBlock: Label, var exceptionType: JType) : Instruction() {

    override val isControl: Boolean
        get() = true

    override fun name(): String {
        return "TRY"
    }

    override fun opcode(): Int {
        return -1
    }

    @Throws(IOException::class)
    override fun read(wide: Boolean, constants: ArrayList<Constant<*>>, method: Method, labelMaker: (Int) -> Label, pc: Int, `in`: DataInputStream) {
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun write(wide: Boolean, out: DataOutputStream, labelIndexer: (Label) -> Int, constantIndexer: (Constant<*>) -> Int, pc: Int) {
        throw UnsupportedOperationException()
    }

    override fun pushes(): Int {
        return 0
    }

    override fun pops(): Int {
        return 0
    }

    override fun toString(): String {
        return "TRY " + exceptionType.niceName + " -> " + catchBlock.toString()
    }

    override fun fromString(str: String): Instruction {
        error("unsupported")
    }
}
